tests/test_adjoint.py

```diff
@@ -155,12 +161,19 @@ def test_adjoint_J(self, mkey, shape, kernel, space_order, setup_func):
         # Compute initial born perturbation from m - m0
         dm = (solver.model.vp.data**(-2) - model0.vp.data**(-2))
 
-        du = solver.jacobian(dm, vp=model0.vp)[0]
+        if setup_func is viscoacoustic_setup:
+            du = solver.jacobian(dm, vp=model0.vp, qp=model0.qp, b=model0.b)[0]
```

Comment:
@mloubout wonder whether would be useful to extend `.jacobian` (and its siblings) to optionally accept the `model` directly so that they can autonomously and transparently extract all the kwargs they need

... rather than having this sort of branching in user or test code

@nogueirapeterson I'm interested in your thoughts as well

Reply:
Yes definitely, I know we talked about it at some point just was low priority so wasn't done.

Reply:
> @nogueirapeterson I'm interested in your thoughts as well

Yes, I was discussing this with @lauerami and we agreed with you. However, to keep the code standardized, this extension is also required for other models (for example acoustic, TTI...). What is your suggestion?

Reply:
An extension/rework of
https://github.com/devitocodes/devito/blob/1e619c0208ecf5f3817d614cd49b8df89d0beb22/examples/seismic/tti/wavesolver.py#L271 

should generalize to any model I think

Reply:
So, according to your suggestion, I have understood that the extension would be as follows (eg. viscoacoustic wavesolver):

```
def jacobian_adjoint(self, rec, p, pa=None, grad=None, vp=None, qp=None, b=None,
                         v=None, va=None, r=None, checkpointing=False,
                         model=None, **kwargs):
        model = model or self.model
        # Pick vp from model unless explicitly provided
        kwargs.update(model.physical_params(
            vp=vp, qp=qp, b=b)
        )
```
Thus, should we extend it for all `class ...WaveSolver(object):`?

Reply:
I'd say something like 
```
def jacobian_adjoint(self, rec, p, pa=None, grad=None, vp=None, qp=None, b=None,
                         v=None, va=None, r=None, checkpointing=False,
                         model=None, **kwargs):
        # Pick vp from model unless explicitly provided
        kwargs.update(model.physical_params(model=model))
```

Reply:
So, `qp`, `b`, `vp` will not be input parameters or should have a conditional `if model is none`?

I would like to know about it because in `test_gradient.py::TestGradient::test_gradientJ` 

```
for i in range(0, 7):
            # Add the perturbation to the model
            def initializer(data):
                data[:] = np.sqrt(v0.data**2 * v**2 /
                                  ((1 - H[i]) * v**2 + H[i] * v0.data**2))
            vloc = Function(name='vloc', grid=wave.model.grid, space_order=space_order,
                            initializer=initializer)
            # Data for the new model
            d = wave.forward(vp=vloc)[0]
            delta_d = (d.data - rec.data).reshape(-1)
            # First order error F(m0 + hdm) - F(m0)
            error1[i] = np.linalg.norm(delta_d, 1)
            # Second order term F(m0 + hdm) - F(m0) - J dm
            error2[i] = np.linalg.norm(delta_d - H[i] * Jdm.data.reshape(-1), 1)
```
where,

In this case, `vp` is overwritten inside the loop. Then, with this change, should we need to instantiate a `model`?

Reply:
Could have both as input yes or as part of `kwargs` with something like `kwargs.update(model.physical_params(model=model, **kwargs))`

Reply:
Hi @mloubout 

We already did the modifications suggested by you. However, we would like to know your opinion regarding docstrings. Taking TTI example:

```
def forward(self, src=None, rec=None, u=None, v=None, model=None,
                save=False, kernel='centered', **kwargs):
        """
        Forward modelling function that creates the necessary
        data objects for running a forward modelling operator.

        Parameters
        ----------
        geometry : AcquisitionGeometry
            Geometry object that contains the source (SparseTimeFunction) and
            receivers (SparseTimeFunction) and their position.
        u : TimeFunction, optional
            The computed wavefield first component.
        v : TimeFunction, optional
            The computed wavefield second component.
        model : Model, optional
            Object containing the physical parameters.
        vp : Function or float, optional
            The time-constant velocity.
        epsilon : Function or float, optional
            The time-constant first Thomsen parameter.
        delta : Function or float, optional
            The time-constant second Thomsen parameter.
        theta : Function or float, optional
            The time-constant Dip angle (radians).
        phi : Function or float, optional
            The time-constant Azimuth angle (radians).
        save : bool, optional
            Whether or not to save the entire (unrolled) wavefield.
        kernel : str, optional
            Type of discretization, centered or shifted.
```

Should we remove the input parameters `vp`, `epsilon`, `delta`, `theta` and `phi` from docstrings? Do you have any suggestions?